pilot/pkg/proxy/envoy/v2/debug.go

```diff
@@ -319,53 +290,99 @@ func (s *DiscoveryServer) authenticationz(w http.ResponseWriter, req *http.Reque
-			if destConfig != nil {
-				rule := destConfig.Spec.(*networking.DestinationRule)
-				clientProtocol = clientAuthProtocol(rule, p)
-			} else {
-				clientProtocol = authnHTTP
-			}
-			info.ClientProtocol = authProtocolToString(clientProtocol)
+			info = append(info, AnalyzeMTLSSettings(ss.Hostname, p, authnPolicy, destConfig)...)
+		}
+	}
 
-			if (clientProtocol & serverProtocol) == 0 {
-				if clientProtocol == authnCustomMTls && serverProtocol != authnHTTP {
-					info.TLSConflictStatus = "MAY CONFLICT"
-				} else {
-					info.TLSConflictStatus = "CONFLICT"
-				}
-			} else {
-				info.TLSConflictStatus = "OK"
-			}
-			if b, err := json.MarshalIndent(info, "  ", "  "); err == nil {
-				_, _ = w.Write(b)
-			}
-			_, _ = fmt.Fprintf(w, ",\n")
+	if b, err := json.MarshalIndent(info, "  ", "  "); err == nil {
+		_, _ = w.Write(b)
+	}
+}
+
+// AnalyzeMTLSSettings returns mTLS compatibility status between client and server policies.
+func AnalyzeMTLSSettings(hostname host.Name, port *model.Port, authnPolicy *authn.Policy,
+	destConfig *model.Config) []*AuthenticationDebug {
+	// TODO(diemvu): add policy config name to the cache push config for this.
+	authnPolicyName := "-"
+	if authnPolicy != nil {
+		authnPolicyName = "???"
+	}
+
+	serverMTLSMode := authn_alpha1.GetMutualTLSMode(authnPolicy)
+
+	baseDebugInfo := AuthenticationDebug{
+		Port:                     port.Port,
+		AuthenticationPolicyName: authnPolicyName,
+		DestinationRuleName:      configName(destConfig),
+		ServerProtocol:           serverMTLSMode.String(),
+		ClientProtocol:           "-",
+	}
+
+	var rule *networking.DestinationRule
+
+	if destConfig != nil {
+		rule = destConfig.Spec.(*networking.DestinationRule)
+	}
+
+	output := []*AuthenticationDebug{}
+
+	clientTLSModes := collectTLSSettingsForPort(rule, port)
+	var subsets []string
+	for k := range clientTLSModes {
+		subsets = append(subsets, k)
+	}
+	sort.Strings(subsets)
+
+	for _, ss := range subsets {
+		c := clientTLSModes[ss]
+		info := baseDebugInfo
+		if c != nil {
+			info.ClientProtocol = c.GetMode().String()
 		}
+		if ss == "" {
+			info.Host = string(hostname)
+		} else {
+			info.Host = fmt.Sprintf("%s|%s", hostname, ss)
+		}
+		info.TLSConflictStatus = EvaluateTLSState(c, serverMTLSMode)
+
+		output = append(output, &info)
 	}
-	_, _ = fmt.Fprint(w, "\n{}]")
+	return output
+}
+
+// EvaluateTLSState returns the conflict state (string) for the input client+server settings.
+// The output string could be:
+// - "AUTO": client TLS settings is not set, and will be automatically determined based on server mode.
+// - "OK": both client and server TLS settings are set correctly.
+// - "CONFLICT": both client and server TLS settings are set, but could be incompatible.
+func EvaluateTLSState(clientMode *networking.TLSSettings, serverMode authn_alpha1.MutualTLSMode) string {
+	const okState string = "OK"
+	const conflictState string = "CONFLICT"
+	const autoState string = "AUTO"
+
+	if clientMode == nil {
```

Comment:
should we return auto even when mesh config does not enable the auto mode?

Reply:
I think the semantic of the mesh "auto mode" is changed slightly. For the `STRICT` mode, we already apply mTLS regardless of this flag (https://github.com/istio/istio/pull/16657). For the permissive, it's just a change from plaintext-as-default to mtls-when-possible. In either case, users don't have to set DR manually but the traffic should still work (though may not be secure).  That's why I use `AUTO`, simply means the DR will be determined by pilot.

Reply:
Customers needs to set DR in PERMISSIVE mode, when mesh config is not enabled. Right?

Reply:
No, they don't. Currently, if user use permissive authN, DR are not added (during installation), and traffic is in plain-text until they add DR to change it to mTLS explicitly.